### ISSUE TITLE
Incorrect display of character 'k' for "FantasqueSansMono" when "ss01" font feature is on

### ISSUE THREAD

**Issue by Frefreak** [ryanoasis/nerd-fonts#901]
#### 🗹 Requirements

- [x] I have searched the [issues](https://github.com/ryanoasis/nerd-fonts/issues) for my issue and found nothing related and/or helpful
- [x] I have searched the [FAQ](https://github.com/ryanoasis/nerd-fonts/wiki/FAQ-and-Troubleshooting) for help
- [x] I have searched the [Wiki](https://github.com/ryanoasis/nerd-fonts/wiki) for help

#### 🎯 Subject of the issue

Experienced behavior:
If a webpage enables "ss01" font feature, and render font is chosen as "FantasqueSansMono Nerd Font", then what should be displayed as
a normal `k` will show like a "coffee" icon. 

Expected behavior:

I see text without emoji

Example symbols:
```
jjjjkkkkllll
```

#### 🔧 Your Setup

- _Which font are you using (e.g. `Anonymice Powerline Nerd Font Complete.ttf`)?_
   - _Please give the full filename_ **Fantasque Sans Mono Regular Nerd Font Complete.ttf**
   - _Where did you get the file from (download link, self patched, source downloaded from link...)_ **archlinux's AUR** [package](https://aur.archlinux.org/packages/nerd-fonts-fantasque-sans-mono)
- _Which terminal emulator are you using (e.g. `iterm2`, `urxvt`, `gnome`, `konsole`)?_  **alacritty (but this issue happens on webpage)**
- _Are you using OS X, Linux or Windows? And which specific version or distribution?_ **archlinux**

#### ★ Screenshots (Optional)

I initially noticed it [here](https://github.blog/2022-08-29-gits-database-internals-i-packed-object-store/):
I override liberation mono and monospace with FantasqueSansMono Nerd Font using fontconfig's config file, so it's very likely what you see will be fine (because it's not rendered as this font). From the devtools image we can see my browser is rendering using "FantasqueSansMono Nerd Font".

![image](https://user-images.githubusercontent.com/7122156/187436836-48c61520-4b5e-4980-ab14-9c314248c40a.png)
![image](https://user-images.githubusercontent.com/7122156/187437578-d3033cc1-b66d-4fb0-b618-924fa798a200.png)

I believe this can be easily reproduced using this html snippet:
```
<style>
* {
  font-feature-settings: "ss01" on;
  font-family: FantasqueSansMono Nerd Font;
}
</style>
<pre><code>
jjjjkkkkllll
</code></pre>
```

![image](https://user-images.githubusercontent.com/7122156/187437831-c65eb40c-2784-48d7-871f-c2b69f51a287.png)

**Comment by Finii** [comment]
Can not reproduce...

![image](https://user-images.githubusercontent.com/16012374/187441159-a4e45ed8-1624-4238-9ff3-cc3df94ef8ed.png)

Ah, SS01 you say, hold tight ...

![image](https://user-images.githubusercontent.com/16012374/187442166-69c308f1-8538-4297-8b5c-8a4131309366.png)

... examining ...

**Comment by Finii** [comment]
Source font is ok

![image](https://user-images.githubusercontent.com/16012374/187443037-79e9693c-2e72-434f-8819-1ef27710d6c8.png)

**Comment by Finii** [comment]
While this is easy to fix in `fontforge` interactively, it seems the necessary actions are not exposed to the Python interface.
What I need is a way to access (edit) the current `SS01` subtable.
I can show it etc, but I can not access (show) the individual rules. I can just add new rules.

### Any hint welcome

Other than that, started investigation to use `fonttools`, but I would rather avoid that. Maybe as postprocess script. Hmm.

**Comment by Finii** [comment]
> Maybe as postprocess script

Ah yes, of course _post_ is too late, it must be _pre_, but we do not have the 'infrastructure' for pre-stuff yet :roll_eyes:

**Comment by Finii** [comment]
This is my 'stub' where I try to get access (at least read access) to the sub table:

```python
for filename in sys.argv[1:]:
    fullfile = os.path.basename(filename)
    fname = os.path.splitext(fullfile)[0]

    font = fontforge.open(filename, 1)
    for tab in font.gsub_lookups:
        r = re.match('[\'"][sS][sS]([0-9][0-9])[\'"].*', tab)
        if not r:
            continue
        print('TABS {}'.format(r.groups()[0]))
        print('     : {}'.format(font.getLookupInfo(tab)))
        print('     : {}'.format(font.getLookupInfo(tab)[2]))
        print('     {}'.format(font.lookupSetFeatureList(tab,font.getLookupInfo(tab)[2])))
        for subtable in font.getLookupSubtables(tab):
            print('   - {}'.format(subtable))
            print('     {}'.format(font.getLookupSubtableAnchorClasses(subtable)))
    font.close()
```

**Comment by Frefreak** [comment]
Thank you for your work.
How font works under the hood is totally new to me so I'm afraid I can't really help anything for now. However I think this might be a good chance to learn those stuffs so I can understand your script and how font patching works. I'm starting with fontforge...

**Comment by Frefreak** [comment]
I now know some basic usage of `fontforge` and have briefly read the `font-patcher` script.

To my understanding the **pomicon** icons should not be moved otherwise when other projects use one of the pomicons it will display incorrectly. Since it looks like `GSUB` table specifies substitution by name (e.g. **k** -> **k.noloop** when ss01 is on),  does we need to programmatically determine there's a **k** to **k.noloop** and move **k.noloop** somewhere else?  Is this understanding right?

I tried all the getXXX functions and just like you said none of them seems to do  what we want to get info like this:
![image](https://user-images.githubusercontent.com/7122156/190197931-1586fc62-abbd-4bd3-ad64-1a7dfa2d9287.png)

Currently the only way I found is through the `glyph` api like so:
![image](https://user-images.githubusercontent.com/7122156/190198697-306c6695-b208-4857-b362-cda830af3796.png)

This would means the script needs to iterate all the glyphs to determine all the mappings (given no prior knowledge), which sounds super inefficient and I don't think should be the way to go. Or can we hard-code the mapping for known fonts, something similar to how src font attrs is doing [here](https://github.com/ryanoasis/nerd-fonts/blob/a07648b1eef52c87670fcd5b567c55493c0b3205/font-patcher#L695-L730)?

**Comment by Finii** [comment]
Thank you for working on it!!

Well, with #914 we already iterate through 'all' glyphs to find refs. Would be easy enough to also look the posSub tables up. And then not patch the ~referenced~ substitution destination glyph.

Iirc I tried to move the ref'ed glyph to some free space and adapt the references, but ... well that seemed a lot of work for very few ref glyphs.
So I choose to skip patching that glyph.

Once you start editing the tables it's a can of worms 😒 as they can be extremely complex.

Font based data files are already here, the `config. cfg` in each individual source font directory. That does not help for self patching though.

And/or `--careful` could get an optional parameter _where_ to be careful. That can be filled by self patcher persons and config.cfg here in the CI.

But is it really an option to skip some symbols on patching just because they are needed for SS12, which probably noone uses? Maybe SS users are required to burn in the style set with a tool (which you need to do anyhow on all the not-styleset-aware terminals ... and then self patch?

These are tough questions.

Just noticed that the (maybe more important?) PPEM fix #761 does not work with `.ttc` #783, of course, and that will also consume lots of time. Sigh.

**Comment by Finii** [comment]
Did not implement the "move to free space" way, but the glyphs will automatically be marked with `careful`, i.e. not replaced.
